File: jsh/loader/loader.js
//	LICENSE
//	This Source Code Form is subject to the terms of the Mozilla Public License, v. 2.0. If a copy of the MPL was not
//	distributed with this file, You can obtain one at http://mozilla.org/MPL/2.0/.
//
//
//	The Original Code is the jsh JavaScript/Java shell.
//
//	The Initial Developer of the Original Code is David P. Caldwell <david@davidpcaldwell.com>.
//	Portions created by the Initial Developer are Copyright (C) 2014 the Initial Developer. All Rights Reserved.
//
//	Contributor(s):
//	END LICENSE

(function() {
	//	TODO	naming conventions are inconsistent in this stuff; look at how there are addClasses methods and classpath.add().
	//			generally speaking, should probably match the rhinoLoader API across all of these representations of it

	//	TODO	The name prefix used below is duplicative of the one in js/debug/plugin.jsh.js, so not DRY currently
	var _log = function(_logger,_level,mask) {
		var substitutions = Array.prototype.slice.call(arguments,3);
		if (_logger.isLoggable(_level)) {
			var _message = Packages.java.lang.String.format(mask, substitutions);
			_logger.log(_level, _message);
		}
	}

	$host.$api.deprecate.warning = function(o) {
		var name = arguments.callee.javaLogName;
		var _level = Packages.java.util.logging.Level.WARNING;
		var _logger = Packages.java.util.logging.Logger.getLogger(name);
		var _traceLevel = Packages.java.util.logging.Level.FINE;
		if (o.callee) {
			if (o.object && o.property) {
				_log(_logger, _level, "Use of deprecated method %s of object %s", String(o.property), String(o.object));
			} else {
				_log(_logger, _level, "Use of deprecated function %s", String(o.callee));
			}
		} else if (o.object && o.property) {
			_log(_logger, _level, "Access to deprecated property %s of object %s", String(o.property), String(o.object));
		}
		if (_logger.isLoggable(_traceLevel)) {
			_log(_logger, _traceLevel, "Stack trace of deprecated usage:\n%s", String(new Error().stack));
		}
		debugger;
	};
	$host.$api.deprecate.warning.javaLogName = "inonit.script.jsh.Shell.log.$api.deprecate";
//	var rhinoLoader = $host;

//	this.bootstrap = function(path,context) {
//		var loader = new rhinoLoader.Loader({
//			_code: $host.loader.getBootstrapModule(path)
//		});
//		return loader.module("module.js", context);
//	}

	var getCode = function(code) {
		if (typeof(code) == "undefined") throw new TypeError("'code' must not be undefined.");
		if (code === null) throw new TypeError("'code' must not be null.");
		//	This check determines whether the object is a Pathname; is there a way to do that in the rhino/file module itself?
		//	TODO	presumably the run/file methods should only support file objects, not directories or pathnames not
		//			corresponding to files ... or else what should they do if the file is not found? Maybe file could return
		//			null or something ... but run would probably have to fail silently, which is not good unless it is
		//			explicitly specified
		if (code.java && code.java.adapt() && $host.classpath.getClass("java.io.File").isInstance(code.java.adapt())) {
			return {
				name: code.toString(),
				_in: new Packages.java.io.FileInputStream(code.java.adapt())
			};
		} else {
			return code;
		}
	}

	return new function() {
		this.plugin = new function() {
			this.read = function(_code,scope) {
				var loader = new $host.Loader({ _source: _code.getScripts() });
				return loader.run("plugin.jsh.js", scope);
			};
			this.run = function(_code,path,scope,target) {
				$host.run(
					{
						_source: _code.getScripts(),
						path: path
					},
					scope,
					target
				);
			};
			this.file = function(_code,path,context) {
				return $host.file(
					{
						_source: _code.getScripts(),
						path: path
					},
					context
				);
			};
			this.module = function(_code,main,context) {
				var loader = new $host.Loader({ _code: _code });
				return loader.module(main, context);
			};
			this._stream = function(_code,path) {
				var _codeSourceFile = _code.getScripts().getFile(path);
				if (_codeSourceFile) return _codeSourceFile.getInputStream();
				return null;
			};
			this.addClasses = function(_code) {
				$host.classpath.add(_code.getClasses());
			}
		}

		this.run = function(code,scope,target) {
			return $host.run(getCode(code),scope,target);
		}

		this.file = function(code,$context) {
			return $host.file(getCode(code),$context);
		}

		this.module = function(pathname) {
			var format = {};
			if (pathname.directory) {
				if (!pathname.directory.getFile("module.js")) {
					return null;
				}
				format.base = pathname.java.adapt();
				format.name = "module.js";
			} else if (pathname.file && /\.slime$/.test(pathname.basename)) {
				format.slime = pathname.java.adapt();
				format.name = "module.js";
			} else if (pathname.file) {
				format.base = pathname.parent.java.adapt();
				format.name = pathname.basename;
			} else {
				return null;
			}
			var p = {};
			if (arguments.length == 2) {
				p.$context = arguments[1];
			}
			var loader = (function(format) {
				if (format.slime) return new $host.Loader({ _packed: format.slime });
				if (format.base) return new $host.Loader({ _unpacked: format.base });
				throw new TypeError("Unreachable code: format.slime and format.base null in jsh loader's module()");
			})(format);
			var args = [format.name].concat(Array.prototype.slice.call(arguments,1));
			return loader.module.apply(loader,args);
		}

		this.classpath = new function() {
			this.toString = function() {
				return $host.classpath.toString();
			}

			this.add = function(_file) {
				$host.classpath.add(Packages.inonit.script.engine.Code.Source.create(_file));
			};

			this.get = function(name) {
				return $host.classpath.getClass(name);
			}
		};

		this.namespace = function(name) {
			return $host.namespace(name);
		}
	}
})()

File: loader/rhino/java/inonit/script/engine/Loader.java
//	LICENSE
//	This Source Code Form is subject to the terms of the Mozilla Public License, v. 2.0. If a copy of the MPL was not
//	distributed with this file, You can obtain one at http://mozilla.org/MPL/2.0/.
//
//	The Original Code is the SLIME loader for rhino.
//
//	The Initial Developer of the Original Code is David P. Caldwell <david@davidpcaldwell.com>.
//	Portions created by the Initial Developer are Copyright (C) 2012-2013 the Initial Developer. All Rights Reserved.
//
//	Contributor(s):
//	END LICENSE

package inonit.script.engine;

import java.io.*;
import java.util.*;
import java.net.*;
import java.util.logging.*;

public abstract class Loader {
	private static final Logger LOG = Logger.getLogger(Loader.class.getName());

	public abstract String getCoffeeScript() throws IOException;
	public abstract String getLoaderCode(String path) throws IOException;

	public static abstract class Classes {
		public static abstract class Configuration {
			public abstract boolean canCreateClassLoaders();
			public abstract ClassLoader getApplicationClassLoader();
			public abstract File getLocalClassCache();
		}
		
		private static HashMap<ClassLoader,Code.Source> cache = new HashMap<ClassLoader,Code.Source>();
		
		private static Code.Source adapt(ClassLoader parent) {
			if (parent instanceof URLClassLoader) {
				List<URL> urls = Arrays.asList(((URLClassLoader)parent).getURLs());
				List<Code.Source> sources = new ArrayList<Code.Source>();
				for (URL url : urls) {
					if (url.getProtocol().equals("file")) {
						try {
							File file = new File(url.toURI());
							if (file.getName().endsWith(".jar")) {
								sources.add(Code.Source.zip(file));
							} else {
								sources.add(Code.Source.create(file));
							}
						} catch (java.net.URISyntaxException e) {
							throw new RuntimeException(e);
						}
					} else {
						sources.add(Code.Source.create(url));
					}
				}
				return Code.Source.create(sources);
			} else {
				return null;
			}
		}

		public static abstract class Interface {
			public abstract void append(Code.Source code);

			abstract Code.Source dependencies();
			abstract ClassLoader classLoader();
			
			/**
				Returns class with the given name, or <code>null</code> if there is no such class.
			*/
			public final Class getClass(String name) {
				try {
					return classLoader().loadClass(name);
				} catch (ClassNotFoundException e) {
					return null;
				}
			}


			public final void append(Code code) {
				append(code.getClasses());
			}
			
			private Code.Source parent;
			
			final Code.Source parent() {
				if (parent == null) {
					parent = adapt(classLoader().getParent());
				}
				return parent;
			}
			
			public final Code unpacked(File base) {
				return Code.loadUnpacked(base, this);
			}
			
			public final Code unpacked(URL base) {
				return Code.loadUnpacked(base, this);
			}
			
			public static final Interface NULL = new Interface() {
				@Override
				public void append(Code.Source code) {
					throw new UnsupportedOperationException("Not supported yet."); //To change body of generated methods, choose Tools | Templates.
				}

				@Override
				Code.Source dependencies() {
					return Code.Source.NULL;
				}
				
				ClassLoader classLoader() {
					return null;
				}
			};
		}

		public abstract ClassLoader getApplicationClassLoader();
		public abstract Interface getInterface();

		public static Classes create(final Configuration configuration) {
			if (configuration.canCreateClassLoaders()) {
				final ClassLoaderImpl loaderClasses = ClassLoaderImpl.create(configuration.getApplicationClassLoader());
				return new Classes() {
					@Override public ClassLoader getApplicationClassLoader() {
						return loaderClasses;
					}

					@Override public Interface getInterface() {
						return loaderClasses.toInterface();
					}
				};
			} else {
				final ClassLoader loader = configuration.getApplicationClassLoader();
				return new Classes() {
					@Override public ClassLoader getApplicationClassLoader() {
						return loader;
					}

					@Override public Interface getInterface() {
						return null;
					}
				};
			}
		}

		private static class ClassLoaderImpl extends ClassLoader {
			static ClassLoaderImpl create(ClassLoader delegate) {
				LOG.log(Level.FINE, "Creating Loader.Classes: parent=%s", delegate);
				return new ClassLoaderImpl(delegate);
			}

			private inonit.script.runtime.io.Streams streams = new inonit.script.runtime.io.Streams();
			private ArrayList<Code.Source> locations = new ArrayList<Code.Source>();

			private ClassLoaderImpl(ClassLoader delegate) {
				super(delegate);
			}

			public String toString() {
				String rv = getClass().getName() + ": delegate=" + this.getParent() + " locations=[";
				synchronized(locations) {
					for (int i=0; i<locations.size(); i++) {
						rv += locations.get(i);
						if (i+1 != locations.size()) {
							rv += ",";
						}
					}
				}
				rv += "]";
				return rv;
			}

			protected Class findClass(String name) throws ClassNotFoundException {
				String path = name.replace('.', '/') + ".class";
				String[] tokens = name.split("\\.");
				String packageName = tokens[0];
				for (int i=1; i<tokens.length-1; i++) {
					packageName += "." + tokens[i];
				}
				synchronized(locations) {
					for (Code.Source source : locations) {
						try {
							Code.Source.File in = source.getFile(path);
							if (in != null) {
								if (getPackage(packageName) == null) {
									definePackage(packageName,null,null,null,null,null,null,null);
								}
								byte[] b = streams.readBytes(in.getInputStream());
								return defineClass(name, b, 0, b.length);
							}
						} catch (IOException e) {
							//	do nothing
						}
					}
				}
				throw new ClassNotFoundException(name);
			}

			protected URL findResource(String name) {
				synchronized(locations) {
					for (Code.Source source : locations) {
						Code.Classes classes = source.getClasses();
						if (classes != null) {
							URL url = classes.getResource(name);
							if (url != null) {
								return url;
							}
						}
					}
				}
				return null;
			}

			protected Enumeration<URL> findResources(String name) {
				java.util.Vector<URL> rv = new java.util.Vector<URL>();
				synchronized(locations) {
					for (Code.Source source : locations) {
						Code.Classes classes = source.getClasses();
						if (classes != null) {
							URL url = classes.getResource(name);
							if (url != null) {
								rv.add(url);
							}
						}
					}
				}
				return rv.elements();
			}
			
			private Classes.Interface api = new Classes.Interface() {
				private Code.Source dependencies = Code.Source.create(locations);
				
				@Override public String toString() {
					return "Loader.Classes.Interface for: " + ClassLoaderImpl.this.toString();
				}

				@Override public void append(Code.Source code) {
					synchronized(locations) {
						locations.add(code);
					}
				}

				Code.Source dependencies() {
					return dependencies;
				}
				
				ClassLoader classLoader() {
					return ClassLoaderImpl.this;
				}
			};

			Classes.Interface toInterface() {
				return api;
			}
		}
	}
}